babel/dates.py

```diff
@@ -1453,10 +1452,44 @@ def format_day_of_year(self, num):
     def format_day_of_week_in_month(self):
         return '%d' % ((self.value.day - 1) // 7 + 1)
 
-    def format_period(self, char):
-        period = {0: 'am', 1: 'pm'}[int(self.value.hour >= 12)]
-        for width in ('wide', 'narrow', 'abbreviated'):
-            period_names = get_period_names(context='format', width=width, locale=self.locale)
+    def format_period(self, char, num):
+        """
+        Return period from parsed datetime according to format pattern.
+
+        >>> format = DateTimeFormat(time(19, 42), 'en_US')
+        >>> format.format_period('a', 1)
+        u'PM'
+
+        >>> format.format_period('b', 1)
+        u'evening'
+
+        >>> format.format_period('B', 1)
+        u'in the evening'
+
+        >>> format = DateTimeFormat(datetime(2022, 4, 28, 6, 27), 'zh_Hant')
+        >>> format.format_period('a', 1)
+        u'上午'
+
+        >>> format.format_period('b', 1)
+        u'清晨'
+
+        >>> format.format_period('B', 1)
+        u'清晨'
+
+        :param char: pattern format character ('a', 'b', 'B')
+        :param num: count of format character
+
+        """
+        widths = [{3: 'abbreviated', 4: 'wide', 5: 'narrow'}[max(3, num)],
+                  'wide', 'narrow', 'abbreviated']
```

Comment:
Just so I'm following, the idea here is to take a preferred width according to `num`, followed by fallbacks?

🤔  Passing in a `num` >= 6 will crash here, but that's fine – those are out of spec anyway.

Reply:
> Just so I'm following, the idea here is to take a preferred width according to `num`, followed by fallbacks?

Yes. But the fallback is not needed if all widths of `dayPeriod` in CLDR are defined.

I'll check CLDR database.

> 🤔 Passing in a `num` >= 6 will crash here, but that's fine – those are out of spec anyway.

Other `format_*` methods have the same issue. It is defined to allow 1-5 characters for `b` and `B` in `PATTERN_CHARS`. Actually, `BBBBBB` (num: 6) is not allowed.

```pycon
>>> dates.format_time(time(13), 'BBBBBB h:mm', locale='en_US')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jun66j5/src/babel.git/babel/dates.py", line 833, in format_time
    return parse_pattern(format).apply(time, locale)
  File "/home/jun66j5/src/babel.git/babel/dates.py", line 1667, in parse_pattern
    raise ValueError('Invalid length for field: %r'
ValueError: Invalid length for field: 'BBBBBB'
```